### PR TITLE
Remove ko.mapping

### LINKED ISSUE #2117: [dev] Remove knockout-mapping
We only use knockout-mapping.js in one module (comment.js). I propose removing it. If need be, we can easily add a helper function to transform a JS object of simple types into an object who values are observables.

### PR DESCRIPTION
**Pull request by Ghalko**
Close #2117 

## Proposed
ko mapping is no longer actively supported. Simple function koHelpers.mapJStoKO does a one-deep mapping observables pass, just passing through arrays and objects.

## Repercussions
In-house means in-house supported. Will need to be made "deep" with a recursive function. Need to add arrayObservable when someone uses it to make an array that needs to be seen to change on the browser.

### CODE REVIEW

File: website/static/js/koHelpers.js [hunk 1]
```diff
@@ -69,6 +69,27 @@ var sanitizedObservable = function(value) {
     });
 };
 
+//Maps an object two deep into a given object as ko observables.
+var mapJStoKO = function(data, parent, array, deep) {
+    array = typeof array !== 'undefined' ? array : false; //Default no observableArray
+    deep = typeof deep !== 'undefined' ? deep : false; //Default top level no recursion
+    //TODO: Remove as implemented.
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Remove this for now.

File: website/static/js/tests/contribManager.test.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+/*global describe, it, expect, example, before, after, beforeEach, afterEach, mocha, sinon*/
```
Comment (sloria) [inline website/static/js/tests/contribManager.test.js]: Remove this, as there are no actual tests here.

File: website/static/js/koHelpers.js [hunk 3]
```diff
@@ -69,6 +69,27 @@ var sanitizedObservable = function(value) {
     });
 };
 
+//Maps an object two deep into a given object as ko observables.
+var mapJStoKO = function(data, parent, array, deep) {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: Rather than mutating `parent`, it would be simpler if this function just took an object as input and another object (with observables) as output. We could then do:

```js
$.extend(self, mapJStoKO(data));
```

File: website/static/js/koHelpers.js [hunk 4]
```diff
@@ -69,6 +69,27 @@ var sanitizedObservable = function(value) {
...
+    array = typeof array !== 'undefined' ? array : false; //Default no observableArray
+    deep = typeof deep !== 'undefined' ? deep : false; //Default top level no recursion
+    //TODO: Remove as implemented.
+    if (array || deep) {
+        console.log("Not implemented yet");
+    }
+    for (var key in data) {
+        if (data.hasOwnProperty(key)) {
+            if (typeof data[key] !== 'object') {
```
Comment (sloria) [inline website/static/js/koHelpers.js]: If the value `data[key]` is an array, it should be an `observableArray` in the output.

Comment (sloria) [inline website/static/js/koHelpers.js]: If `data[key]` is an `Object`, why not just make it an observable instead of skipping it?

File: website/static/js/koHelpers.js [hunk 5]
```diff
@@ -69,6 +69,27 @@ var sanitizedObservable = function(value) {
     });
 };
 
+//Maps an object two deep into a given object as ko observables.
```
Comment (sloria) [inline website/static/js/koHelpers.js]: This goes one-deep, not two deep

File: website/static/js/tests/koHelpers.test.js [hunk 6]
```diff
@@ -18,4 +20,42 @@ describe('koHelpers', () => {
...
+            };
+            koHelpers.mapJStoKO(data, parent);
+        });
+
+        it('makes observables', () => {
+            assert.isTrue(ko.isObservable(parent.thursday));
+            assert.isTrue(ko.isObservable(parent.more));
+            assert.isTrue(ko.isObservable(parent.number));
+            assert.isTrue(parent.hasOwnProperty("object2"));
+            assert.isTrue(parent.hasOwnProperty("array2"));
```
Comment (sloria) [inline website/static/js/tests/koHelpers.test.js]: Test that arrays are turned into `observableArrays` (see previous comment)

**Comment by sloria** [comment]
There's a lot of noise in the commit history for this PR. @Ghalko , can you send a new PR with just your changes (and the unnecessary files removed, as per my comments)?

File: website/static/js/koHelpers.js (hunk 4 above)
Comment (Ghalko) [inline website/static/js/koHelpers.js]: Not every array needs to be observable. Not every object needs to be observable.

Comment (sloria) [inline website/static/js/koHelpers.js]: That is true for any type of variable (not all `Numbers` and `Strings` need to be observables; why special-case `Arrays` and `Objects`?

Ideally, you'd have a way to exclude variables. something like

```js
mapJStoKO({foo: 42, bar: ['baz', 'qux']}, {exclude: ['bar']});
```

**Comment by samchrisinger** [comment]
I worry that any automatic mapping pattern is going to create problems down the line. From my brief experiences with ko.mapping I found it has the potential to make code difficult to follow. For example when I replace
```
self.nodeHasAuth(data.nodeHasAuth)
self.userCanWrite(data.userCanWrite)
etc ...
```
with
```
koHelpers.mapJStoKO(data, self);
```
The code loses its capacity for being self-documenting. Perhaps we can include docstrings around each usage of mapJStoKO, but then we're not saving any typing and are losing flexibility. 

Additionally, any automatic mapping pattern also makes implicit assumptions about the ko type the incoming data is cast to. Not all arrays need be observable. Not all objects need to have their members make observable, and sometimes we'll want observables whose values are objects. 

Am not opposed to ever using this pattern, but think it needs to be used thoughtfully.

**Comment by sloria** [comment]
Valid point, @samchrisinger . The original thought was to replace a large plugin (which we only used once) with a simple function. Perhaps the better solution is to always be explicit when defining observables.

Thoughts, @Ghalko ?

**Comment by Ghalko** [comment]
With the new one that will become a PR soon:

```
var mapJStoKO = function(data, exclude) {
```

Allows the user to exclude the things that don't need to be observable. But without python's 

```
def mapSuchandSuch(data=something, exclude=None):
```

It isn't very self documenting. So the default use of this would be to make everything observable from the passed in data. And if we make it recursive, how do you specify one more down? Then you have to compile a list.

Just an idea but what if default behavior is to make everything observable, but have include and exclude options. Essentially for a big dataset that needs most excluded you can pass in a list of things that need mapping. And vice versa?

**Comment by sloria** [comment]
I think what @samchrisinger is suggesting is different. 

Automatic mapping can make code hard to follow because it's not immediately apparent what instance variables are initialized in a ViewModel's constructor.

```js
this.title = ko.observable('default title');
this.enabled = ko.observable(false);
$.extend(this, mapJStoKO(data));

// much later in the code
this.content()  // where did this come from? what was its initial value?
```

It's possible we don't need a mapping function/plugin at all, and we always explicitly define observables.

**Comment by samchrisinger** [comment]
Right. Another alternative might be to explicitly define observables in the VM constructor (with defaults or nulls), and then use some function as a generic way to update that VM (i.e. from some post-instantiation async call). 

But this approach still doesn't address difficulties in nesting, and it's very possible the types specified in the constructor will get clobber by the update call, which might be even worse than undocumented observables.
